com.unity.multiplayer.mlapi/Runtime/Core/MessageBatcher.cs

```diff
@@ -0,0 +1,198 @@
+                ProfilerStatManager.rpcsSent.Record();
+            }
+        }
+
+        public delegate void SendCallbackType(ulong clientId, SendStream messageStream);
+        public delegate void ReceiveCallbackType(BitStream messageStream, MLAPI.RPCQueueManager.QueueItemType messageType, ulong clientId, float time);
+
+        /// <summary>
+        /// SendItems
+        /// Send any batch of RPC that are of length above threshold
+        /// </summary>
+        /// <param name="threshold"> the threshold in bytes</param>
+        /// <param name="sendCallback"> the function to call for sending the batch</param>
+        public void SendItems(int threshold, SendCallbackType sendCallback)
+        {
+            List<ulong> sent = new List<ulong>();
+
+            foreach (KeyValuePair<ulong, SendStream> entry in SendDict)
+            {
+                // read the queued message
+                int length = (int)SendDict[entry.Key].Stream.Length;
+
+                if (length >= threshold)
+                {
+                    sendCallback(entry.Key, entry.Value);
+                    ProfilerStatManager.rpcBatchesSent.Record();
+
+                    // mark the client for which a batch was sent
+                    sent.Add(entry.Key);
+                }
+            }
+
+            // clear the batches that were sent from the SendDict
+            // this cannot be done above acuring the Dictionary iteration, so we do it here
+            foreach(ulong clientid in sent)
+            {
+                SendDict.Remove(clientid);
+            }
+        }
+
+        /// <summary>
+        /// ReceiveItems
+        /// Process the messageStream and call the callback with individual RPC messages
+        /// </summary>
+        /// <param name="messageStream"> the messageStream containing the batched RPC</param>
+        /// <param name="receiveCallback"> the callback to call has type int f(message, type, clientId, time) </param>
+        /// <param name="messageType"> the message type to pass back to callback</param>
+        /// <param name="clientId"> the clientId to pass back to callback</param>
+        /// <param name="receiveTime"> the packet receive time to pass back to callback</param>
+        public int ReceiveItems(in BitStream messageStream, ReceiveCallbackType receiveCallback, MLAPI.RPCQueueManager.QueueItemType messageType, ulong clientId, float receiveTime)
+        {
+            do
+            {
+                // read the length of the next RPC
+                int rpcSize = PopLength(messageStream);
+
+                // copy what comes after current stream position
+                long pos = messageStream.Position;
+                BitStream copy = new BitStream();
+                copy.SetLength(rpcSize);
+                copy.Position = 0;
+                Buffer.BlockCopy(messageStream.GetBuffer(), (int)pos, copy.GetBuffer(), 0, rpcSize);
```

Comment:
I hope we'll get to a point where we don't copy but indicate a pre-allocated memory block with a Slice/Span...